ilastik/experimental/parser/_base.py

```diff
@@ -33,44 +31,50 @@ def __exit__(self, exc_type, exc, tb):
 
 def _create_project_wrap(hdf5_file):
     type_ = hdf5_file[WORKFLOW_KEY][()]
-    if type_ == PIXEL_CLASSIFICATION:
-        return _PixelClassProjectImpl(hdf5_file)
-
-    raise NotImplementedError("Unknown project type {type_}")
-
-
-class _Keys:
-    INPUT_DATA = "Input Data"
-    INPUT_DATA_INFOS = "infos"
-    INPUT_DATA_RAW = "Raw Data"
-    INPUT_DATA_AXIS_TAGS = "axistags"
-    INPUT_DATA_SHAPE = "shape"
-
-    FEATURES = "FeatureSelections"
-    FEATURES_IDS = "FeatureIds"
-    FEATURES_SCALES = "Scales"
-    FEATURES_COMPUTE_IN_2D = "ComputeIn2d"
-    FEATURES_SELECTION_MATRIX = "SelectionMatrix"
-    PIXEL_CLASSIFICATION = "PixelClassification"
-    PIXEL_CLASSIFICATION_TYPE = "pickled_type"
-    PIXEL_CLASSIFICATION_FACTORY = "ClassifierFactory"
-    PIXEL_CLASSIFICATION_FORESTS = "ClassifierForests"
+    try:
+        return types.registry[type_](hdf5_file)
+    except KeyError:
+        raise NotImplementedError(f"Unknown project type {type_}")
+
+
+class _InputDataKeys:
+    ROOT = "Input Data"
+    INFOS = "infos"
+    RAW = "Raw Data"
+    AXIS_TAGS = "axistags"
+    SHAPE = "shape"
+
+
+class _PixelFeatureKeys:
+    ROOT = "FeatureSelections"
+    IDS = "FeatureIds"
+    SCALES = "Scales"
+    COMPUTE_IN_2D = "ComputeIn2d"
+    SELECTION_MATRIX = "SelectionMatrix"
+
+
+class _PixelClassificationKeys:
+    ROOT = "PixelClassification"
+    TYPE = "pickled_type"
+    FACTORY = "ClassifierFactory"
+    FORESTS = "ClassifierForests"
     LABEL_NAMES = "LabelNames"
 
 
-class _PixelClassProjectImpl(types.PixelClassificationProject):
```

Comment:
Any reason for holding onto HDF5 file handle and not parsing all data into a dataclass? This way, there is no need for ABCs to exist.

```python
@dataclass
class Project:
    _registry: ClassVar[MutableMapping[str, "Project"]] = {}

    def __init_subclass__(cls, workflow_name: str = "", **kwargs):
        if workflow_name:
            Project._registry[workflow_name] = cls

    common_prop_1: ...
    common_prop_2: ...

    @classmethod
    def _from_ilp(cls, f: h5py.Group):
        return cls(
            common_prop_1=f[...],
            common_prop_2=f[...],
            ...
        )


@dataclass
class ObjectClassificationProject(Project):
    common_for_oc_prop: ...

    @classmethod
    def _from_ilp(cls, f: h5py.Group):
       return cls(
           common_for_oc_prop=f[...],
           **dataclasses.asdict(super()._from_ilp(f)),
       )


@dataclass
class PredictionsObjectClassificationProject(ObjectClassificationProject, workflow_name="Object Classification (from prediction image)"):
    ...


def from_ilp(path):
    with h5py.File(path, "r") as f:
        workflow_name = f[...]
        try:
            return Project._registry[workflow_name]._from_ilp(f)
       except KeyError:
            raise ValueError(f"unsupported workflow {workflow_name}") from None
```